torchvision/prototype/datasets/utils/_resource.py

```diff
@@ -1,51 +1,171 @@
+import hashlib
+import itertools
 import pathlib
-from typing import Optional, Union
+import warnings
+from typing import Optional, Sequence, Tuple, Callable, IO, Any, Union, NoReturn
 from urllib.parse import urlparse
 
-from torch.utils.data import IterDataPipe
-from torch.utils.data.datapipes.iter import IterableWrapper
-from torchdata.datapipes.iter import IoPathFileLoader
+from torchdata.datapipes.iter import (
+    IoPathFileLoader,
+    IterableWrapper,
+    FileLister,
+    IterDataPipe,
+    ZipArchiveReader,
+    TarArchiveReader,
+    RarArchiveLoader,
+)
+from torchvision.datasets.utils import (
+    download_url,
+    _detect_file_type,
+    extract_archive,
+    _decompress,
+    download_file_from_google_drive,
+)
 
 
-# FIXME
-def compute_sha256(path: pathlib.Path) -> str:
-    return ""
+class OnlineResource(abc.ABC):
+    def __init__(
+        self,
+        *,
+        file_name: str,
+        sha256: Optional[str] = None,
+        decompress: bool = False,
+        extract: bool = False,
+        preprocess: Optional[Callable[[pathlib.Path], pathlib.Path]] = None,
+        loader: Optional[Callable[[pathlib.Path], IterDataPipe[Tuple[str, IO]]]] = None,
+    ) -> None:
+        self.file_name = file_name
+        self.sha256 = sha256
 
+        if preprocess and (decompress or extract):
+            warnings.warn("The parameters 'decompress' and 'extract' are ignored when 'preprocess' is passed.")
+        elif extract:
+            preprocess = self._extract
+        elif decompress:
+            preprocess = self._decompress
+        self._preprocess = preprocess
 
-class LocalResource:
-    def __init__(self, path: Union[str, pathlib.Path], *, sha256: Optional[str] = None) -> None:
-        self.path = pathlib.Path(path).expanduser().resolve()
-        self.file_name = self.path.name
-        self.sha256 = sha256 or compute_sha256(self.path)
+        if loader is None:
+            loader = self._default_loader
+        self._loader = loader
 
-    def to_datapipe(self) -> IterDataPipe:
-        return IoPathFileLoader(IterableWrapper((str(self.path),)), mode="rb")  # type: ignore
+    @staticmethod
+    def _extract(file: pathlib.Path) -> pathlib.Path:
+        return pathlib.Path(
+            extract_archive(str(file), to_path=str(file).replace("".join(file.suffixes), ""), remove_finished=False)
+        )
 
+    @staticmethod
+    def _decompress(file: pathlib.Path) -> pathlib.Path:
+        return pathlib.Path(_decompress(str(file), remove_finished=True))
 
-class OnlineResource:
-    def __init__(self, url: str, *, sha256: str, file_name: str) -> None:
-        self.url = url
-        self.sha256 = sha256
-        self.file_name = file_name
+    def _default_loader(self, path: pathlib.Path) -> IterDataPipe[Tuple[str, IO]]:
+        if path.is_dir():
+            return IoPathFileLoader(FileLister(str(path), recursive=True), mode="rb")
+
+        dp = IoPathFileLoader(IterableWrapper((str(path),)), mode="rb")
```

Comment:
FYI there is currently no benefits of using `IoPathFileLoader` for remote downloading, as `Pathlib` doesn't support urls. The rest of the implementation doesn't necessarily support this use-case either so not blocking, but good to keep in mind as we might want to move away from Pathlib for this reason